### DISCUSSION TITLE
How to use transaction through multiple services?

### DISCUSSION THREAD

**Discussion by maxinmos** [benbjohnson/wtf#22] [Q&A] [answered]
Hello, thanks for great series.
But currently, you use transaction in every query inside services. So how to use transaction through multiple services?
Example:
```
tx, _ := d.Begin()

// store file
fileStorageService.SaveImage("/avatar.jpg")

// database stuffs
userService.CreateUser(user)

// other database stuffs
...

tx.Commit()
```

**Comment by benbjohnson** [comment, marked as answer]
Good question. I typically make my service methods implement a transactional boundary so that I don't need to share the transaction outside of it. In the example you have, you don't have a shared transaction as the `fileStorageService` isn't likely transactional (assuming it's disk-backed or S3-backed). You can end up with the file on your storage service but your `CreateUser()` could fail.

You could model your `Tx` inside your application domain but that assumes that all your services that use `myapp.Tx` are transactional. Transactions really only make sense within a single service (e.g. a `sqlite` layer) as you can't easily implement transactions across multiple layers or implementations.

**Reply by maxinmos** [reply]
Thanks very much for your answer. 
How do you think about "transaction per request" pattern (https://goenning.net/2017/06/20/session-per-request-pattern-go).
In .NET Entity Framework, DbContext is created per request, but I don't know why it's not common in Go.

**Reply by benbjohnson** [reply]
I personally dislike having the HTTP layer know too much about the underlying data access. Typically, I structure my HTTP handlers to only handle the transport (e.g. encoding/decoding) and that just hands off to the service implementation.

For example, in the WTF Dial repo the data access for HTTP handlers is pretty minimal:

- https://github.com/benbjohnson/wtf/blob/daadc79f3778fd49db6e4064878030487e2e2a47/http/dial.go#L63-L68
- https://github.com/benbjohnson/wtf/blob/daadc79f3778fd49db6e4064878030487e2e2a47/http/dial.go#L117-L129
- https://github.com/benbjohnson/wtf/blob/daadc79f3778fd49db6e4064878030487e2e2a47/http/dial.go#L171-L172
- https://github.com/benbjohnson/wtf/blob/daadc79f3778fd49db6e4064878030487e2e2a47/http/dial.go#L218-L223
- https://github.com/benbjohnson/wtf/blob/daadc79f3778fd49db6e4064878030487e2e2a47/http/dial.go#L246-L255
- https://github.com/benbjohnson/wtf/blob/daadc79f3778fd49db6e4064878030487e2e2a47/http/dial.go#L274-L278
- https://github.com/benbjohnson/wtf/blob/daadc79f3778fd49db6e4064878030487e2e2a47/http/dial.go#L307-L311

**Reply by maxinmos** [reply]
I see, thanks a lot.
